src/MacVim/MMCoreTextView.m

```diff
@@ -1708,529 +1417,121 @@ - (void)batchDrawData:(NSData *)data
-
-    CGContextSaveGState(context);
-
-    int originalFontSmoothingStyle = 0;
-    if (thinStrokes) {
-        CGContextSetShouldSmoothFonts(context, YES);
-        originalFontSmoothingStyle = CGContextGetFontSmoothingStyle(context);
-        CGContextSetFontSmoothingStyle(context, fontSmoothingStyleLight);
-    }
-
-    // NOTE!  'cells' is zero if we're drawing a composing character
-    CGFloat clipWidth = cells > 0 ? cells*cellSize.width : w;
-    CGRect clipRect = { {x, y}, {clipWidth, cellSize.height} };
-    CGContextClipToRect(context, clipRect);
-
-    if (!(flags & DRAW_TRANSP)) {
-        // Draw the background of the text.  Note that if we ignore the
-        // DRAW_TRANSP flag and always draw the background, then the insert
-        // mode cursor is drawn over.
-        CGRect rect = { {x, y}, {cells*cellSize.width, cellSize.height} };
-        CGContextSetRGBFillColor(context, RED(bg), GREEN(bg), BLUE(bg),
-                                 ALPHA(bg));
-
-        // Antialiasing may cause bleeding effects which are highly undesirable
-        // when clearing the background (this code is also called to draw the
-        // cursor sometimes) so disable it temporarily.
-        CGContextSetShouldAntialias(context, NO);
-        CGContextSetBlendMode(context, kCGBlendModeCopy);
-        CGContextFillRect(context, rect);
-        CGContextSetShouldAntialias(context, antialias);
-        CGContextSetBlendMode(context, kCGBlendModeNormal);
-    }
-
-    if (flags & DRAW_UNDERL) {
-        // Draw underline
-        CGRect rect = { {x, y+0.4*fontDescent}, {cells*cellSize.width, 1} };
-        CGContextSetRGBFillColor(context, RED(sp), GREEN(sp), BLUE(sp),
-                                 ALPHA(sp));
-        CGContextFillRect(context, rect);
-    } else if (flags & DRAW_UNDERC) {
-        // Draw curly underline
-        int k;
-        float x0 = x, y0 = y+1, w = cellSize.width, h = 0.5*fontDescent;
-
-        CGContextMoveToPoint(context, x0, y0);
-        for (k = 0; k < cells; ++k) {
-            CGContextAddCurveToPoint(context, x0+0.25*w, y0, x0+0.25*w, y0+h,
-                                     x0+0.5*w, y0+h);
-            CGContextAddCurveToPoint(context, x0+0.75*w, y0+h, x0+0.75*w, y0,
-                                     x0+w, y0);
-            x0 += w;
+    __block size_t cells_filled = 0;
+    [string enumerateSubstringsInRange:NSMakeRange(0, string.length)
+                               options:NSStringEnumerationByComposedCharacterSequences
+                            usingBlock:^(NSString *substring, NSRange substringRange, NSRange enclosingRange, BOOL *stop) {
+        GridCell *cell = grid_cell_safe(&grid, row, col + cells_filled++ * (wide ? 2 : 1));
+        GridCellInsertionPoint insertionPoint = {0};
+        if (flags & DRAW_TRANSP)
+            insertionPoint = cell->insertionPoint;
+        if ([substring isEqualToString:@" "]) {
+            *cell = (GridCell){ .bg = bg, .insertionPoint = insertionPoint };
```

Comment:
Need to set the other variables such as `fg`/`flags`/`sp` as well to make e.g. underline draw properly.

Reply:
Done.